File: leetcode/cpp/shortest-path-with-alternating-colors.cpp
/* -------------------------------------------------------------------------- */
/*     https://leetcode.com/problems/shortest-path-with-alternating-colors    */
/* -------------------------------------------------------------------------- */

#include <iostream>
#include <queue>
#include <vector>

#include "common-struct.h"

using namespace std;

class Solver {
   private:
    int num_nodes_;
    vector<vector<int>> red_graph_;
    vector<vector<int>> blue_graph_;

    vector<int> distance_via_blue_;
    vector<int> distance_via_red_;

    enum Color { Red, Blue };

   public:
    Solver(int n, vector<vector<int>>& red_edges, vector<vector<int>>& blue_edges) {
        num_nodes_ = n;
        red_graph_.resize(n);
        blue_graph_.resize(n);
        distance_via_blue_.resize(num_nodes_, INT_MAX);
        distance_via_red_.resize(num_nodes_, INT_MAX);
        for (const auto& edge : red_edges) {
            red_graph_[edge[0]].push_back(edge[1]);
        }

        for (const auto& edge : blue_edges) {
            blue_graph_[edge[0]].push_back(edge[1]);
        }
    }

    void bfs() {
        queue<pair<int, Color>> q;

        // Assume you reached 0th node via blue edge
        q.push({0, Color::Blue});
        distance_via_blue_[0] = 0;

        // Assume you reached 0th node via red edge
        q.push({0, Color::Red});
        distance_via_red_[0] = 0;

        while (!q.empty()) {
            const auto [node, came_via] = q.front();
            q.pop();

            if (came_via == Color::Blue) {
                int dist_for_nbr = distance_via_blue_[node] + 1;

                // You'll have to go nbrs via red edge
                for (const auto nbr : red_graph_[node]) {
                    if (dist_for_nbr < distance_via_red_[nbr]) {
                        q.push({nbr, Color::Red});
                        distance_via_red_[nbr] = dist_for_nbr;
                    }
                }
            } else if (came_via == Color::Red) {
                int dist_for_nbr = distance_via_red_[node] + 1;

                // You'll have to go nbrs via blue edge
                for (const auto nbr : blue_graph_[node]) {
                    if (dist_for_nbr < distance_via_blue_[nbr]) {
                        q.push({nbr, Color::Blue});
                        distance_via_blue_[nbr] = dist_for_nbr;
                    }
                }
            }
        }
    }

    vector<int> solve() {
        bfs();
        vector<int> ans(num_nodes_, -1);
        for (int i = 0; i < num_nodes_; i++) {
            int dist = min(distance_via_blue_[i], distance_via_red_[i]);
            if (dist != INT_MAX) {
                ans[i] = dist;
            }
        }
        return ans;
    }
};

int main() {
    vector<vector<int>> red = {{0, 1}};
    vector<vector<int>> blue = {{1, 2}};
    PrintVector(Solver(3, red, blue).solve());
}

File: leetcode/golang/src/shortest-path-with-alternating-colors.go
/* -------------------------------------------------------------------------- */
/*     https://leetcode.com/problems/shortest-path-with-alternating-colors    */
/* -------------------------------------------------------------------------- */

package main

import "math"

type Color int

const (
	Blue Color = iota
	Red
)

type Pair struct {
	node    int
	cameVia Color
}

func getGraphs(n int, redEdges [][]int, blueEdges [][]int) ([][]int, [][]int) {
	redGraph := make([][]int, n)
	for i := range redGraph {
		redGraph[i] = []int{}
	}

	blueGraph := make([][]int, n)
	for i := range blueGraph {
		blueGraph[i] = []int{}
	}

	for _, edge := range redEdges {
		redGraph[edge[0]] = append(redGraph[edge[0]], edge[1])
	}

	for _, edge := range blueEdges {
		blueGraph[edge[0]] = append(blueGraph[edge[0]], edge[1])
	}

	return redGraph, blueGraph
}

func bfs(redGraph, blueGraph [][]int, distanceViaRed, distanceViaBlue []int) {
	queue := []Pair{}
	for i := range distanceViaBlue {
		distanceViaBlue[i] = math.MaxInt32
		distanceViaRed[i] = math.MaxInt32
	}
	queue = append(queue, Pair{0, Blue})
	queue = append(queue, Pair{0, Red})

	distanceViaBlue[0] = 0
	distanceViaRed[0] = 0

	for len(queue) > 0 {
		front := queue[0]
		queue = queue[1:]

		if front.cameVia == Red {
			nbr_new_dist := distanceViaRed[front.node] + 1

			for _, nbr := range blueGraph[front.node] {
				if nbr_new_dist < distanceViaBlue[nbr] {
					queue = append(queue, Pair{nbr, Blue})
					distanceViaBlue[nbr] = nbr_new_dist
				}
			}
		}

		if front.cameVia == Blue {
			nbr_new_dist := distanceViaBlue[front.node] + 1
			for _, nbr := range redGraph[front.node] {
				if nbr_new_dist < distanceViaRed[nbr] {
					queue = append(queue, Pair{nbr, Red})
					distanceViaRed[nbr] = nbr_new_dist
				}
			}
		}
	}

}

func ShortestAlternatingPaths(n int, redEdges [][]int, blueEdges [][]int) []int {

	redGraph, blueGraph := getGraphs(n, redEdges, blueEdges)

	distanceViaBlue := make([]int, n)
	distanceViaRed := make([]int, n)
	bfs(redGraph, blueGraph, distanceViaBlue, distanceViaRed)

	distance := []int{}
	for i := range distanceViaBlue {
		dist := distanceViaBlue[i]
		if distanceViaRed[i] < dist {
			dist = distanceViaRed[i]
		}

		if dist == math.MaxInt32 {
			dist = -1
		}
		distance = append(distance, dist)
	}

	return distance
}
